scripts: refactor gceworker for managed-service support

ibreakthecloud opened the pull request:
GCEWorker script has been used widely by devs to spawn a remote gcp vm configured to develop CRDB.
In order to work with managed-service in the same fashion, this script is hacked or modified. This commit adds
the support of managed-service in gceworker script out-of-the-box and eliminates the need to either spin new
vm or change the script.
Changes:
- adds support for two new env var DISABLE_COCKROACHDB & DISABLE_MANAGED_SERVICE
  toggling anyone to true will NOT configure them.
- adds cloning for managed-service in bootstrap-debian.sh script
- adds installation of neccessary deps and env for managed-service

Default behaviour of script is to install and configure enviroment for both the repo, Cockroach & managed-service.

Note: Devs still needs to manually do a git clone of managed-service in go paths

Part of: DEVINF-1371
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/139469)
<!-- Reviewable:end -->

rickystewart commented on build/bootstrap/bootstrap-debian.sh:
```diff
@@ -70,3 +72,25 @@ sudo mv /tmp/bazelisk /usr/bin/bazel
+# If DISABLE_MANAGED_SERVICE is false, clone Managed Service
+if [ "${DISABLE_MANAGED_SERVICE:-false}" = "false" ]; then
+  git clone git@github.com:cockroachlabs/managed-service.git "$(go env GOPATH)/src/github.com/cockroachlabs/managed-service"
+  git -C "$(go env GOPATH)/src/github.com/cockroachlabs/managed-service" submodule update --init
```
This repo appears to have no submodules. Am I missing something?

ibreakthecloud replied:
correcting it.

rickystewart commented on build/bootstrap/bootstrap-debian.sh:
```diff
@@ -27,18 +27,14 @@ sudo apt-get install -y --no-install-recommends \
   nodejs \
-  bison
+  bison \
+  protobuf-compiler
```
Why this change? The `README` [says](https://github.com/cockroachlabs/managed-service#install-bazel) that you should NOT have `protoc` installed globally.

ibreakthecloud replied:
If I am not wrong, this is for mac machine. Right above [this line](https://github.com/cockroachlabs/managed-service?tab=readme-ov-file#install-google-cloud-sdk) readme says to install globally using `brew`.

rickystewart replied:
The `README` clearly says that having `protoc` globally installed breaks the build. 

1. Why do you want to install it?
2. If you're going to go against what the `README` says, did you at least test whether what it says is accurate?

rickystewart replied:
I don't believe it needs to be installed globally.

ibreakthecloud replied:
I have removed this, and pushed commit with all the suggested changes. In the background I will be checking if skipping `protobuf` is giving any issue with cockroach and managed-service development.

rickystewart commented on scripts/gceworker.sh:
```diff
@@ -99,7 +102,15 @@ case "${cmd}" in
+    gcloud compute ssh "${NAME}" --ssh-flag="-A" --command="DISABLE_COCKROACHDB=$DISABLE_COCKROACHDB DISABLE_MANAGED_SERVICE=$DISABLE_MANAGED_SERVICE ./bootstrap/bootstrap-debian.sh"
+
+    # Copy the ssh keys to the gceworker, this is needed to clone the managed-service repo
+    echo -n "Enter the path to the ssh key [default ~/.ssh]: "
```
Making the message clearer would be an improvement. What SSH key am I supposed to use? What will it be used for?

What if I don't want to copy any of my SSH keys? If I don't want to clone `managed-service`, copying it isn't necessary.

rickystewart replied:
Given the message, people will assume that you can input something like `~` and it will be expanded to the home directory. Is that correct? Did you verify that that works?

ibreakthecloud replied:
`~` will not expand to home dir. I am updating the text. Thanks for pointing out.

rickystewart commented on scripts/gceworker.sh:
```diff
@@ -99,7 +102,15 @@ case "${cmd}" in
+    if [ -z "$SSH_KEY_PATH" ]; then
+      SSH_KEY_PATH="$HOME/.ssh"
+    fi
+    gcloud compute scp --recurse "${SSH_KEY_PATH}" "${NAME}:~/.ssh"
```
Do you not need both the private and public keys?

How does this work given that you already called `bootstrap-debian.sh` above? Wouldn't cloning the repo have already failed?

ibreakthecloud replied:
> How does this work given that you already called bootstrap-debian.sh above? Wouldn't cloning the repo have already failed?

fixing this

rickystewart commented on build/bootstrap/bootstrap-debian.sh:
```diff
@@ -6,8 +6,8 @@
 #
-# On a Debian/Ubuntu system, bootstraps a docker install and the cockroach
-# repo.
+# On a Debian/Ubuntu system, bootstraps all the requried dependencies for the
```
required

rickystewart commented on build/bootstrap/bootstrap-debian.sh:
```diff
@@ -70,3 +72,25 @@ sudo mv /tmp/bazelisk /usr/bin/bazel
 . bootstrap/bootstrap-unison.sh
+
+# Configure environment variables for CockroachDB and Managed Service.
+echo 'export PATH="${PATH}:$HOME/go/src/github.com/cockroachdb/cockroach/bin:/usr/local/go/bin"' >> ~/.bashrc_bootstrap
```
It seems you should only be doing this stuff if you are cloning the repo, right? If you don't clone the repo, there's no benefit to updating the `PATH` like this.

ibreakthecloud replied:
That's a valid point. I kept it there because seemed harmless at first. I'll move this.

rickystewart commented on build/bootstrap/bootstrap-debian.sh:
```diff
@@ -55,9 +51,15 @@ sha256sum -c - <<EOF
+docker-compose --version
+
+# Install NVM.
+# Note: you still required to run `nvm install <version>` to install a specific version of Node.js.
```
We already install `node` above (line 29). Why is this necessary?

ibreakthecloud replied:
If user need to develop code using a different version of Node, nvm is really useful in switching node version. Although L29 installation of `node` can be removed.

rickystewart replied:
Did you check what version of `node` is installed and whether you would want a different version?

ibreakthecloud replied:
The current node version is `v16.20.2`. And managed-service I beleive [needs](https://github.com/cockroachlabs/managed-service/blob/master/console/ui/cc-console/package.json#L184C15-L184C22) `20.10.0` or above. Hence `nvm` would be useful for someone developing managed-service ui.

jlinder commented on scripts/gceworker.sh:
```diff
@@ -14,6 +14,9 @@ source build/shlib.sh
 export CLOUDSDK_CORE_PROJECT=${CLOUDSDK_CORE_PROJECT-${GCEWORKER_PROJECT-cockroach-workers}}
 export CLOUDSDK_COMPUTE_ZONE=${GCEWORKER_ZONE-${CLOUDSDK_COMPUTE_ZONE-us-east1-b}}
 
+DISABLE_COCKROACHDB=${DISABLE_COCKROACHDB-false}
+DISABLE_MANAGED_SERVICE=${DISABLE_MANAGED_SERVICE-false}
```
Question. Is this a case of [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it)? Is anyone actually going to use this?

What if in this PR just always installs everything and, if people say it now takes too long to spin up a new VM, then we add in this feature (or the appropriate feature given feedback)? This way, if no one uses this, then we won't need to maintain it and if people _do_ ask for something else because it's taking too long to spin up a VM, then we can build the thing they are asking for.

ibreakthecloud replied:
Probably yes! 
My intention of having this feature was not to optimise the vm spin time. It was rather done to avoid burdening vm with additional files and tools, which, I agree, is also not something bothering.
I will remove this for now and will wait for someone to actually ask for this.

jlinder commented on scripts/gceworker.sh:
```diff
@@ -99,7 +102,15 @@ case "${cmd}" in
+    if [ -z "$SSH_KEY_PATH" ]; then
+      SSH_KEY_PATH="$HOME/.ssh"
+    fi
+    gcloud compute scp --recurse "${SSH_KEY_PATH}" "${NAME}:~/.ssh"
```
This looks like it will grab _all_ ssh keys, not just one. What does it grab?

If it grabs all the keys, that is way more than should be pushed up. 

How about making the default explicitly be the common default key of `${HOME}/.ssh/id_rsa`?

jlinder reviewed (changes requested):
Here are a few more things to consider.

ibreakthecloud committed:
scripts: refactor gceworker for managed-service support

GCEWorker script has been used widely by devs to
spawn a remote gcp vm configured to develop CRDB.
In order to work with managed-service in the same fashion,
this script is hacked or modified. This commit adds
the support of managed-service in gceworker script
out-of-the-box and eliminates the need to either spin new
vm or change the script.
Changes:
- adds installation of neccessary deps and env for managed-service

Default behaviour of script is to install and configure
enviroment for both the repo, Cockroach & managed-service.

Note: Devs still needs to manually do a git clone of
managed-service in go paths

Part of: DEVINF-1371
Release note: None

rickystewart commented on scripts/gceworker.sh:
```diff
@@ -98,6 +111,12 @@ case "${cmd}" in
+    # Note: ssh-keyscan is needed to avoid the "Host key verification failed" error 
+    # when cloning the managed-service repo non-interactively
+    gcloud compute ssh "${NAME}" --command="mkdir -p ~/.ssh && ssh-keyscan github.com >> ~/.ssh/known_hosts"
+    gcloud compute scp "${SSH_KEY_FILE_PATH}" "${NAME}:~/.ssh/"
```
Don't you additionally need the public key?

ibreakthecloud replied:
No. AFAIK, public key is not really required in gceworker to clone or push.

jlinder commented on build/bootstrap/bootstrap-debian.sh:
```diff
@@ -34,11 +34,6 @@ curl -fsSL https://get.pnpm.io/install.sh | env PNPM_VERSION=8.6.6 sh -
 
-# Configure environment variables.
-echo 'export PATH="${PATH}:$HOME/go/src/github.com/cockroachdb/cockroach/bin:/usr/local/go/bin"' >> ~/.bashrc_bootstrap
-echo '. ~/.bashrc_bootstrap' >> ~/.bashrc
-. ~/.bashrc_bootstrap
```
Make sure to rebase given #139469 that modified this code. You'll want to incorporate that change in this PR too.

ibreakthecloud replied:
Yeah, there are expected merge conflicts, I will resolve them.

ibreakthecloud replied:
Rebased and fixed merge conflicts.

jlinder commented on build/bootstrap/bootstrap-debian.sh:
```diff
@@ -55,9 +50,15 @@ sha256sum -c - <<EOF
-git clone https://github.com/cockroachdb/cockroach "$(go env GOPATH)/src/github.com/cockroachdb/cockroach"
-git -C "$(go env GOPATH)/src/github.com/cockroachdb/cockroach" submodule update --init
+# Install Docker compose.
+sudo curl -L "https://github.com/docker/compose/releases/download/v2.31.0/docker-compose-linux-x86_64" -o /usr/local/bin/docker-compose
```
Is there a reason we want to install docker-compose this way rather than [using the .deb package for ubuntu](https://phoenixnap.com/kb/install-docker-compose-on-ubuntu)?

ibreakthecloud replied:
This installation steps came from Jay's note of using gceworker for managed-service. Also, using [this method](https://phoenixnap.com/kb/install-docker-compose-on-ubuntu), installs docker compose as a plugin in `docker` binary rather than installing a new binary, `docker-compose`. Usage will be `docker compose` rather `docker-compose`. 
If you refer the code block right above [this](https://github.com/cockroachlabs/managed-service?tab=readme-ov-file#contributing-to-frontend-code) in readme, it say `docker-compose up --build generate`.

I am okay to change though.

ibreakthecloud replied:
I don't think ubuntu comes pre-bundled with `docker-compose`. We do installation of `docker.io` that also installs `docker` cli along with docker engine.
I am keeping this change to install `docker-compose`, in future while upgrading we will move to docker plugin like `docker compose`.

jlinder replied:
Keeping the installation of `docker-compose` like you wrote it is OK. Though, today I found there is a `docker-compose-v2` package available to install. It's currently at version 2.24.6, which is about a year old. It's likely current enough for our needs. Perhaps consider using that instead?

Also, I should have been clearer in my last comment. 

I checked (by starting up a gceworker and seeing what is available to install) that `docker-compose` is available on ubuntu to just install using the package manager. At the time, I only saw the 1.x version in the package list. So I was intending to make clear that the way it's being installed in this PR right now was necessary.

Additionally, our systems require that `docker-compose` be installed because it's used in the code in various places (`docker compose` is not used). So installing the plugin isn't necessary and likely won't be necessary in the future either.

ibreakthecloud replied:
`apt install docker-compose-v2` install `compose` as plugin, `docker compose`.
Keeping this unchanged to install the Docker Compose standalone.

I will change this URL to `https://github.com/docker/compose/releases/latest/download/docker-compose-linux-x86_64` to ensure it always installs the latest `docker-compose`.

One other way to do this, which IMO would be overkill is:
- Install `docker compose` (plugin) using Ubuntu package manager.
- Create a symlink of `/usr/local/bin/docker-compose` to `/usr/libexec/docker/cli-plugins/docker-compose`
This will ensure installation of both plugin and standalone.

jlinder commented on build/bootstrap/bootstrap-debian.sh:
```diff
@@ -70,3 +71,14 @@ sudo mv /tmp/bazelisk /usr/bin/bazel
+git clone https://github.com/cockroachdb/cockroach "$(go env GOPATH)/src/github.com/cockroachdb/cockroach"
+git -C "$(go env GOPATH)/src/github.com/cockroachdb/cockroach" submodule update --init
+
+git clone git@github.com:cockroachlabs/managed-service.git "$(go env GOPATH)/src/github.com/cockroachlabs/managed-service"
```
If the name of the file is _not_ `id_rsa` for the private key (or another standard key name), then cloning the managed-service repo will fail as written. Perhaps make a link to the file using the `id_rsa` name if the chosen filename is different?

ibreakthecloud replied:
I am now creating a symlink of copied key to `~/.ssh/id_rsa`

jlinder commented on scripts/gceworker.sh:
```diff
@@ -74,6 +74,19 @@ case "${cmd}" in
       read COCKROACH_DEV_LICENSE
     fi
 
+    DEFAULT_SSH_KEY_PATH="$HOME/.ssh/id_rsa"
```
I suggest always using curly braces when using variables in shell scripts. This helps avoid problems. This also matches the usage in the rest of the scripts. If you make this change, then please change it for all your code changes. 

```suggestion
    DEFAULT_SSH_KEY_PATH="${HOME}/.ssh/id_rsa"
```

jlinder reviewed (changes requested):
A few more comments.

ibreakthecloud commented on build/bootstrap/bootstrap-debian.sh:
```diff
@@ -70,3 +71,14 @@ sudo mv /tmp/bazelisk /usr/bin/bazel
+git clone https://github.com/cockroachdb/cockroach "$(go env GOPATH)/src/github.com/cockroachdb/cockroach"
+git -C "$(go env GOPATH)/src/github.com/cockroachdb/cockroach" submodule update --init
+
+git clone git@github.com:cockroachlabs/managed-service.git "$(go env GOPATH)/src/github.com/cockroachlabs/managed-service"
```
Done

jlinder commented on build/bootstrap/bootstrap-debian.sh:
```diff
@@ -72,3 +72,15 @@ sudo mv /tmp/bazelisk /usr/bin/bazel
 . bootstrap/bootstrap-unison.sh
+
+# Configure environment variables for CockroachDB and Managed Service.
+echo 'export PATH="${PATH}:$HOME/go/src/github.com/cockroachdb/cockroach/bin:/usr/local/go/bin"' >> ~/.bashrc_bootstrap
+echo 'export PATH="${PATH}:$HOME/go/src/github.com/cockroachlabs/managed-service/bin"' >> ~/.bashrc_bootstrap
```
```suggestion
echo 'export PATH="${PATH}:${HOME}/go/src/github.com/cockroachdb/cockroach/bin:/usr/local/go/bin"' >> ~/.bashrc_bootstrap
echo 'export PATH="${PATH}:${HOME}/go/src/github.com/cockroachlabs/managed-service/bin"' >> ~/.bashrc_bootstrap
```

ibreakthecloud replied:
Oops! missed this.

jlinder reviewed (commented):
A few comments.

ibreakthecloud commented:
@jlinder I have re-pushed the commit with suggested changes. PTAL.

jlinder commented on build/bootstrap/bootstrap-debian.sh:
```diff
@@ -72,3 +72,15 @@ sudo mv /tmp/bazelisk /usr/bin/bazel
+# Configure environment variables for CockroachDB and Managed Service.
+echo 'export PATH="${PATH}:${HOME}/go/src/github.com/cockroachdb/cockroach/bin:/usr/local/go/bin"' >> ~/.bashrc_bootstrap
+echo 'export PATH="${PATH}:${HOME}/go/src/github.com/cockroachlabs/managed-service/bin"' >> ~/.bashrc_bootstrap
+echo -e '\n' >> ~/.bashrc
```
I saw you changed the other instance of this line to `echo >> ~/.bashrc`. Just wanted to note this one in case you intended to keep them in sync.

ibreakthecloud replied:
Sorry. Missed this!

jlinder commented on build/bootstrap/bootstrap-debian.sh:
```diff
@@ -57,9 +51,15 @@ sha256sum -c - <<EOF
-git clone https://github.com/cockroachdb/cockroach "$(go env GOPATH)/src/github.com/cockroachdb/cockroach"
-git -C "$(go env GOPATH)/src/github.com/cockroachdb/cockroach" submodule update --init
+# Install Docker compose.
+sudo curl -L "https://github.com/docker/compose/releases/latest/download/docker-compose-linux-x86_64" -o /usr/local/bin/docker-compose
```
Let's either revert this to a pinned version or switch to installing `docker-compose-v2` with `apt-get`. We want to always use an exact, known version. This makes running the same command multiple times do the same thing (vs. having something change out from under you from one run to the next).

ibreakthecloud replied:
> switch to installing docker-compose-v2 with apt-get
this installs `compose` as a plugin rather than standalone. I am switching to pinned version.

ibreakthecloud replied:
@jlinder updated the PR. PTAL.

jlinder reviewed (changes requested):
One last change.

jlinder reviewed (approved):
:shipit: !!!

ibreakthecloud commented:
TFTR @jlinder ! 🚀

ibreakthecloud commented:
bors r+
